### PR TITLE
Use branch head reference

### PR DESCRIPTION
**Pull request by hoshin**

### CODE REVIEW

File: controllers/suggestionsController.js [hunk 1]
```diff
@@ -21,42 +20,36 @@ class SuggestionsController {
...
-								}
-							});
-						} else {
-							self.renderErrorPage(res);
-						}
-					});
-				} else {
-					self.renderErrorPage(res);
-				}
+			const pullRequestFlowPromise = githubHelper.getHeadReferenceForBranch('master');
```
Comment (MattiSG) [inline controllers/suggestionsController.js]: The target should be configurable.

File: controllers/suggestionsController.js [hunk 2]
```diff
@@ -21,42 +20,36 @@ class SuggestionsController {
...
+
+			return pullRequestFlowPromise.then( (headerReference) => {
```
Comment (MattiSG) [inline controllers/suggestionsController.js]: > No need for parens with a single param in a fat arrow syntax.

File: controllers/suggestionsController.js [hunk 3]
```diff
@@ -21,42 +20,36 @@ class SuggestionsController {
 	 @param res: An expressjs Response object to get back to the user
 	 */
 	createPullRequest(accessToken, title, description, state, res) {
-		const self = this;
 		const now = (new Date()).getTime();
+		const githubHelper = this.githubHelper();
```
Comment (MattiSG) [inline controllers/suggestionsController.js]: > Use a getter and keep referring to `this.helper`.

File: controllers/suggestionsController.js [hunk 4]
```diff
@@ -21,42 +20,36 @@ class SuggestionsController {
...
-					});
-				} else {
-					self.renderErrorPage(res);
-				}
+			const pullRequestFlowPromise = githubHelper.getHeadReferenceForBranch('master');
+
+			return pullRequestFlowPromise.then( (headerReference) => {
+				return githubHelper.createReference(accessToken, newBranchName, headerReference);
+			})
+			.then( () => {
```
Comment (MattiSG) [inline controllers/suggestionsController.js]: Why the leading space?

File: controllers/suggestionsController.js [hunk 5]
```diff
@@ -21,42 +20,36 @@ class SuggestionsController {
...
+				const testFileName = FILE_NAME_PREFIX + now + '.txt';
```
Comment (MattiSG) [inline controllers/suggestionsController.js]: > Template string.

File: controllers/suggestionsController.js [hunk 6]
```diff
@@ -21,42 +20,36 @@ class SuggestionsController {
...
+				const stateStringBuffer = new Buffer(state);
+				const base64FileContents = stateStringBuffer.toString('base64');
+				return githubHelper.createContent(accessToken, testFileName, newBranchName, description, base64FileContents);
+			})
+			.then( () => {
+				return githubHelper.createPullRequest(newBranchName, title, description, accessToken);
```
Comment (MattiSG) [inline controllers/suggestionsController.js]: > You can also inline this to improve readability and avoid `return` in a fat arrow syntax.

File: helpers/githubHelper.js [hunk 7]
```diff
@@ -1,52 +1,100 @@
 'use strict';
 let superAgent = require('superagent');
-let config = require('../ludwig-conf');
+const githubConfig = {
+	referencesEndpoint:'https://api.github.com/repos/hoshin/git-api-tests/git/refs',
+	createContent:'https://api.github.com/repos/hoshin/git-api-tests/contents/',
+	createPullRequest:'https://api.github.com/repos/hoshin/git-api-tests/pulls'
```
Comment (MattiSG) [inline helpers/githubHelper.js]: Hardcoded.

File: helpers/githubHelper.js [hunk 8]
```diff
@@ -1,52 +1,100 @@
...
-			.post(this.config().github.apiEndpoints.createPullRequest)
-			.send(this.createPullRequestRequestBody(head, title, body))
-			.set('Authorization', `token ${accessToken}`)
-			.end((err, dataToReturn) => {
-				callback(null, dataToReturn);
-			});
+
+	createPullRequest(head, title, body, accessToken) {
+		return new Promise( (resolve) => {
+			this.agent()
```
Comment (MattiSG) [inline helpers/githubHelper.js]: > Now that you've switched to ES6, [`fetch`](https://github.com/sgmap/mes-aides-ui/blob/rewrite/front/cross-browser/fetch.js) would be a pretty good option  :wink:

File: helpers/githubHelper.js [hunk 9]
```diff
@@ -1,52 +1,100 @@
...
+
+	getHeadReferenceForBranch(requestedBranch) {
+		return new Promise( (resolve, reject) => {
+			this.agent()
+				.get(this.config().referencesEndpoint)
+				.end((err, response) => {
+
+					const responseBody = response.body;
+					if (responseBody && Array.isArray(responseBody)) {
+						let branchRef = null;
```
Comment (MattiSG) [inline helpers/githubHelper.js]: Why init to `null`?

File: helpers/githubHelper.js [hunk 10]
```diff
@@ -1,52 +1,100 @@
...
+							if (singleReference.ref === `refs/heads/${requestedBranch}`) {
+								branchRef = singleReference.object.sha;
+							}
+						});
+						if (branchRef) {
+							resolve(branchRef);
+						} else {
+							reject({
+								message: 'Required branch not found',
+								details: `Reference searched for : refs/heads/${requestedBranch}`
```
Comment (MattiSG) [inline helpers/githubHelper.js]: No leading space on `:` in English.

File: helpers/githubHelper.js [hunk 11]
```diff
@@ -1,52 +1,100 @@
...
+							});
+						}
+					} else {
+						reject({message: 'Not able to retrieve references', details: err? err.message : ''});
```
Comment (MattiSG) [inline helpers/githubHelper.js]: `err? err.message : ''` <=> `err && err.message`

File: test/controllers/suggestionsControllerSpec.js [hunk 12]
```diff
@@ -20,27 +20,31 @@ describe('suggestionController', () => {
 			res = {render:sinon.spy()};
 		});
 
-		it('should render the ok page if all remote calls work without errors', () => {
+		it('should render the ok page if all remote calls work without errors', (done) => {
 			//setup
 			const accessToken = 'access token', title = 'title', description = 'description';
+			var res = {render:sinon.spy()};
```
Comment (MattiSG) [inline test/controllers/suggestionsControllerSpec.js]: `var`

File: test/helpers/githubHelperSpec.js [hunk 13]
```diff
@@ -78,11 +83,105 @@ describe('Github Helper', () => {
...
+			const config = [ {
+				pattern: 'https://api.github.com/(.*)',
+				get: () => {
+					throw new Error('Can\'t retrieve references');
+				},
+				fixtures: (match) => {
+					if (match[1] === 'repos/user/reponame/git/refs') {
+						return {};
+					}
+					return {};
```
Comment (MattiSG) [inline test/helpers/githubHelperSpec.js]: Doesn't look like the condition does anything…

Comment (hoshin) [inline test/helpers/githubHelperSpec.js]: This is not needed. Removed in efb45300bca7a867fc904a206668ac6647beeb07

File: controllers/suggestionsController.js (hunk 5 above)
Comment (hoshin) [inline controllers/suggestionsController.js]: Sould be fixed in 621876ee8b03b6bd830b5ec39c5ce1293f760a23

File: helpers/githubHelper.js (hunk 9 above)
Comment (hoshin) [inline helpers/githubHelper.js]: No particular reason appart from old habbits that tend to die hard. Sould be fixed in 621876ee8b03b6bd830b5ec39c5ce1293f760a23

File: helpers/githubHelper.js (hunk 10 above)
Comment (hoshin) [inline helpers/githubHelper.js]: Sould be fixed in 621876ee8b03b6bd830b5ec39c5ce1293f760a23

File: helpers/githubHelper.js (hunk 11 above)
Comment (hoshin) [inline helpers/githubHelper.js]: Sould be fixed in 621876ee8b03b6bd830b5ec39c5ce1293f760a23

File: test/controllers/suggestionsControllerSpec.js (hunk 12 above)
Comment (hoshin) [inline test/controllers/suggestionsControllerSpec.js]: Sould be fixed in 621876ee8b03b6bd830b5ec39c5ce1293f760a23

File: helpers/githubHelper.js (hunk 7 above)
Comment (hoshin) [inline helpers/githubHelper.js]: I thought that's what we agreed upon for github config that'll probably never move, but sure I can extract it in a specific config file =)

File: helpers/githubHelper.js (hunk 11 above)
Comment (MattiSG) [inline helpers/githubHelper.js]: No need to add ref to closing commit, GitHub does it for you  :)

File: helpers/githubHelper.js (hunk 7 above)
Comment (MattiSG) [inline helpers/githubHelper.js]: `hoshin/git-api-tests`? :confused:

File: controllers/suggestionsController.js (hunk 6 above)
Comment (hoshin) [inline controllers/suggestionsController.js]: Commit 1dc437edff8013d8e216745714fcd396fc7e7d77 should take care of this

File: helpers/githubHelper.js (hunk 7 above)
Comment (hoshin) [inline helpers/githubHelper.js]: Oh ... not quite what intended to do, I'll get on this right away

Comment (hoshin) [inline helpers/githubHelper.js]: should be fixed in d32e6dcdec6824b599a5e70d058dae74b2b5a4f9

**Comment by MattiSG** [comment]
Remember to delete merged feature branches  :wink:

**Comment by hoshin** [comment]
Will do, sorry about that ^^'

**Pull request merged**